tests/integ/test_dhcp_exchange.py

```diff
@@ -0,0 +1,37 @@
+from greensim.random import uniform, constant
+
+from itsim.machine.endpoint import Endpoint
+from itsim.network.link import Link
+from itsim.network.router import Router
+from itsim.network.service.dhcp import DHCPDaemon, dhcp_client
+from itsim.simulator import Simulator
+from itsim.types import Protocol, AddressRepr, as_address
+from itsim.units import S, MS, MbPS
+
+
+def set_addresses(*ars: AddressRepr):
+    return {as_address(ar) for ar in ars}
+
+
+def test_dhcp_exchange():
+    sim = Simulator()
+
+    link = Link("10.1.128.0/18", uniform(100 * MS, 200 * MS), constant(100 * MbPS))
+    router = Router().connected_to(link, "10.1.128.1")
+    router.networking_daemon(sim, Protocol.UDP, 67)(DHCPDaemon(100, link.cidr, router.addresses()))
```

Comment:
This syntax is super weird. We should instead go with a function whose last parameter could be a function, or nothing at all. Was that not the point of method `subscribe_networking_daemon`? You find that name too long now, do you? :laughing: 

Also, passing `router.addresses()` to DHCPDaemon is unnecessary: as a Process, it can access these addresses early in its initialization sequence, but later than in its constructor.

```suggestion
    router.networking_daemon(sim, Protocol.UDP, 67, DHCPDaemon(100, link.cidr))
```

Reply:
I don't like this syntax either, I just threw this in to get the test passing. I will either change this to a call to `subscribe_networking_daemon` or revisit the interface soon.

I may be misunderstanding how the addresses are accessed. They were originally collected by a function call when the `Daemon` was first called on, but I removed the call so that the address allocation could be handled at initialization (to remove the requirement for an extra named function and to put all of the validation in `__init__`). Is there a strong reason to wait until the `Daemon` is run to do this allocation?

Reply:
I have been thinking about this and I realize I had it all wrong. The DHCP daemon must be tasked with delivering a *networking configuration* to clients. Such a configuration involves:

1. A unique IP address local to the link
2. A set of routes (`Route` instances) for non-local packets. Typically, a single route is thus provided, in the form of the IP address to the network's *gateway*.
3. A set of DNS servers (which we have got to yet).

I could be wrong, but I'm pretty sure that the `DHCPDaemon` instance looked at its node's IP addresses to figure which is the one local to the link it is configured for, and use that as gateway address to send to the clients. This is much too specific as far as configuration choices go: a DHCP daemon could be set up on any node, not necessarily the router; in that case, the gateway address would still be the router's. So I think a proper set up would include an explicit gateway address, or a `Relay`-class route holding the router's address as gateway. Let me thus propose an alternative change:
```suggestion
    router.networking_daemon(sim, Protocol.UDP, 67)(DHCPDaemon(100, link.cidr, "10.1.128.1"))
```
`DHCPDaemon` would take there an `AddressRepr`. We could have, as an alternative, a sequence of routes, but such a case is not needed at the moment.

Reply:
Also, it would be wise to have that router's address defined as a constant, to avoid litteral litter. (Y

Reply:
Question 1 - The DHCP daemon owns the configuration of the local network: where this or that packet should go to be relayed, and in modern networks (end of '90s), how to find things by name. Thus, it delivers such configuration, either by its DHCPOFFER or DHCPACK packet. I'm not sure which and for our purpose, it's not crucial.

Question 2 - I've run a toy network where the DHCP server daemon ran on another machine than the one relaying packets to the wider area. The network was a point-to-point experiment linking my Raspberry Pi to my own laptop, and the Pi was running DHCP for giggles. It certainly is not a common configuration, so we may not want to hold tight to this sort of flexibility.

Question 3 - A `Relay`-class route is an instance of the `itsim.network.Relay` class, which is a subclass of `itsim.network.Route`. It indicates that a packet whose destination is beyond the local network should be delivered to a certain gateway node, which is then expected to further the packet along its path to its nominal destination.

But these questions are but tangential to the sort of logic that I believe we should actually put in this daemon. Let me start a new comment with this bit of pontification.

Reply:
I think we should go over this tomorrow morning when we're all in the office, since it sounds like there are a lot of aspects of the DHCP implementation that you're expecting to be added or changed that I hadn't fully grasped. It sounds like there's a melange of debates to be had about what criteria define salient (which may be a function of our timeline) and how the details of the implementation should be handled, including, but not limited to, what occurs at configuration time and what occurs at simulation time (which seem to be two subsets of run time that we need to consider). In the meantime, I'll focus on cleaning up the existing code and generalizing it as much as I can so we can quickly make any required changes.